ISPN-6989 Performance guide

slaskawi opened the pull request:
https://issues.jboss.org/browse/ISPN-6989

rvansa commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,178 @@
+
+* `Total Data Set` - Estimated size of all data
+* `Nodes` - The number of nodes in the cluster
+* `Node Failures` - Number of failures possible (also number of owners)
```
number of owners == node failures + 1

slaskawi replied:
Done

rvansa commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,178 @@
+When you don't want to load an entry from a Cache Store, use `Flag.SKIP_CACHE_LOAD` as shown below:
+
+[source,java]
+.Flag.IGNORE_RETURN_VALUES
```
Skip cache load, not ignore return values. However, this can break things.

slaskawi replied:
Done.

Yeah, I've seen the discussion on dev mailing list.

rvansa commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,178 @@
+}
+----
+
+The Externalizer must be registered in cache configuration.
```
example of configuration

slaskawi commented:
Fixed comments. Ready for round #2.

gustavonalle commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,197 @@
+== Number of threads (Server mode only)
+
+Hot Rod Server uses worker threads which are activated by a client's requests.
+It's important to match the number of worker threads to the number of concurrent client requests:
```
The threading in Netty does not follow the thread-per-request model, and many cache operations are executed directly in the EventLoop thread and thus not via worker pool, so not sure if this applies

slaskawi replied:
Hmmm so any suggestions how I should rephrase it?

slaskawi replied:
[JDG7 Perf Guide](https://access.redhat.com/documentation/en-US/Red_Hat_JBoss_Data_Grid/7.0/pdf/Performance_Tuning_Guide/Red_Hat_JBoss_Data_Grid-7.0-Performance_Tuning_Guide-en-US.pdf) says this:

> Worker threads, unlike system threads, are threads activated by a client's request and do not interact
with the user. As the server is asynchronous client requests will continue to be received after this limit
is hit; instead, this number represents the number of active threads performing simultaneous
operations, typically writes.

Maybe we should say it should match the number of concurrent write requests?

gustavonalle replied:
With the latest server changes, AFAICT, not even writes are done in the worker threads unless there is a listener attached to the cache, everything happens in the IO threads that is not configurable afaict (fixed at 2 * processors) so I'd say changing the worker threads size does not make any difference at all.
@wburns WDYT?

galderz replied:
Hmmm, I thought @wburns had changed things so that we don't execute any cache operations in the event loop any more...?

slaskawi replied:
@wburns Any update on this?

slaskawi replied:
From the code I can see that worker threads are used to construct an `Executor`: https://github.com/infinispan/infinispan/blob/master/server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodServer.java#L187-L188

Then we are using this executor to process client requests: https://github.com/infinispan/infinispan/blob/966569022711401369a1a2c33704dd8e5ce2c44f/server/hotrod/src/main/java/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.java#L50-L50 and https://github.com/infinispan/infinispan/blob/966569022711401369a1a2c33704dd8e5ce2c44f/server/hotrod/src/main/java/org/infinispan/server/hotrod/ContextHandler.java#L58-L58

So unless I misinterpret it - this paragraph as it is, makes perfectly sense.

gustavonalle replied:
There's a ```local-handler``` before that processes requests in the IO loop thread to avoid the latency penalty of the executor. I could not find the code that conditionally disable the executor based on the presence/absence of listeners, maybe this was not done in master

gustavonalle commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,197 @@
+
+Infinispan stores data in memory. In case of Local or Replicated mode, all data needs to fit in there,
+so calculating the amount of required memory is trivial.
+
```
It'd be interesting to put some guidelines in order to estimate or calculate the amount of memory. Need to take into account the (serialized or not) key size, (serialized or not) value size, the Infinispan entry overhead (200 bytes?), and the JVM overhead (object header, alignment, etc). This can be done with estimation using tools like http://openjdk.java.net/projects/code-tools/jol/ or actually measuring the heap with a subset of data.

slaskawi replied:
Added proper paragraphs about that.

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+[stem]
+.Total Data Set in library mode
+++++
+Total Data Set = Number Of Entries * (Key Size + Value Size + 200 b (Me tadata))
```
s/Me tadata/Metadata/

How precise is this overhead number? I believe it actually depends a lot on what whether you are using expiration, versioning, the functional API etc.

slaskawi replied:
> s/Me tadata/Metadata/

This is correct :) [Asciimath](http://asciimath.org/) replaces `eta` with `η` :) Don't ask me why they implement it this way...

The overhead numbers are taken from the product documentation (see links in the JIRA). I guess it was estimated or measured.... Maybe @alanfx has some ideas how to put some meaningful information here...

danberindei replied:
I guess it makes some expressions more compact, but I'm guessing the real reason is that they don't split on word boundaries first...

alanfx replied:
Actually, @mgencur has a test to check for this overhead, since it changes from release to release. I do not believe that it looks at all possible combinations of overhead, so maybe @danberindei we should discuss all of the configurations where the overhead should be checked?

slaskawi replied:
@danberindei @alanfx I would also mention only some average values here.... The client needs to be able to roughly estimate the amount of required memory. Calculating this on a byte level makes no sense ;)

danberindei replied:
@slaskawi Sure, we don't need to be too precise, but if we have a test that users can measure the exact overhead with, it would be nice to include it here.

Thinking about this some more, I'd like to change the formula to reference an "Overhead" term, and define it in the paragraph above. I'd call it "Overhead" instead of "Metadata", because a basic embedded cache without versioning and expiration doesn't have any metadata, but it still has an overhead for each entry.

slaskawi replied:
Ok, I changed it to `Overhead` and added a short description.

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+[stem]
+.Total Data Set in server mode
+++++
+Total Data Set = Number Of Entries * (Serialized Key Size + Serialized Value Size + 200 b (Me tadata))
```
Again, versioning and compatibility mode can change the overhead.

slaskawi replied:
I changed phrase:

> and total amount of required memory can be calculated using the following formulas

to 

> and total amount of required memory can be roughly estimated using the following formulas

This should emphasize that this is only an estimate.

danberindei replied:
ok

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+[stem]
+.Required memory in Distributed mode
+++++
+Required Memo ry = (Total Data Set)/(Nodes)*(1+(Node Failures)/(Nodes - Node Failures))
```
I think this should be more like `(Total Data Set) * (Number of Owners + 1) / (Minimum Number of Nodes)`.

Saying minimum number of nodes covers both when nodes crash and when the cluster is scaled down. The `+ 1` is to allow for rebalancing, as during rebalancing some segments will have an extra copy. To be fair, rebalancing also adds some extra overhead, like keeping a list of all the keys that have been modified, so it's hard to estimate accurately.

slaskawi replied:
If I see correctly, the changes you're proposing are:

- Replacing `NodeFailures` with `NumberOfOwners` (as a reminder - `NodeFailures = NumberOfOwners - 1`
- Introducing `MinimumNumberOfNodes`, which is equal to `Nodes - NodeFailures`

I have no strong opinion for the former (I'm OK with either `NumberOfOwners` and `NodeFailures`) but I'm against the latter. Often times when you create a system you're thinking about the possible failures and put VMs (or physical machines) in different racks. So I believe system architects or developers are much more familiar with the number of node failures.

You also changed the equation a little bit, so let me try to do some math on concrete numbers:
```
Total Data Set = 10 GB
Number of Owners = 2
Node Failures = Number of Owners - 1 = 2 - 1 = 1
Nodes = 10
MinumumNumberOfNodes = 10 - 1 = 9

Original formula: (10/10) * (1+(1/(10-1))) = 1 * (1+1/9) = 1.11 GB
Dan's formula: 10 * (2+1) / 9 = 3.33 GB
```

I think the amount of required memory is a bit high and I also don't understand the part with `(number of owners + 1)`.... Could you please elaborate more about this? What is the motivation behind this?

//cc @alanfx @mgencur @rvansa

slaskawi replied:
I'm personally OK with this. However since we are suggesting something else than for the produce I would like to hear @alanfx and @tristantarrant opinion.

alanfx replied:
@danberindei's math makes more sense to me. And to combine it with my recommendation above, you'd actually need 6.67GB or 9.99GB of RAM per node based the intended Infinispan usage.

slaskawi replied:
Sounds good to me... @tristantarrant ?

slaskawi replied:
Assuming @tristantarrant if fine with this... updating...

slaskawi replied:
I changed the formula to: `Required Memory = Total Data Set*(Node Failures + 2)/(Nodes - Node Failures)`. As you can see I preserved node failures since I believe they are more descriptive. Please let me know if you don't like it.

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+Calculated amount of memory should be used for setting `Xmx` and `Xms` parameters.
+
+JVM as well as Infinispan require additional memory for other tasks like searches, allocating network buffers etc.
+It is advised to allocate no more than 50% of memory with living data.
```
@alanfx I think it was even less than 50%...

alanfx replied:
@danberindei it was 50% of the heap when only put, get, and remove are used. More like 30%-33% of the heap when analytics are being used in the cache. (i.e. querying, distributed execution, streams, etc.)

slaskawi replied:
Ok leaving 50% than

alanfx replied:
I'd prefer if the text described both situations! "...no more than 50% of memory with living data when using Infinispan solely as a caching data store, and no more than 33% of memory with living data when using Infinispan to store and analyze the data using querying, distributed execution or distributed streams."?

slaskawi replied:
Done, thanks!

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+JVM as well as Infinispan require additional memory for other tasks like searches, allocating network buffers etc.
+It is advised to allocate no more than 50% of memory with living data.
+
+When considering large heaps, make sure there's enough CPU to perform garbage collection efficiently.
```
I wouldn't mention CPUs, memory bandwidth is usually more important for garbage collection. I'd just mention that long GC pauses can negatively impact the stability of the cluster (in addition to the latency of the application using it), so they should size their heaps accordingly.

slaskawi replied:
I disagree with you here... I would like to avoid people creating VMs (e.g. using OpenStack) with 32GB or RAM and 1 vCPU for example. In that case GC would be terribly slow. 

Also, you often can not change memory bandwidth... But on the other hand you can [measure it](https://specs.openstack.org/openstack/nova-specs/specs/liberty/approved/mem-bw.html).

danberindei replied:
Most of the time, you can also split the 32GB VM into two 16GB VMs. IMO that's usually more effective for reducing the GC pause than adding CPUs.

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+* `-Xmn` - Defines the minimum and maximum value for the young generation.
+* `-XX:NewRatio` - Define the ratio between young and old generations. Should not be used if -Xmn is enabled.
+
+Using `Xms` equal to `Xmx` will prevent JVM from dynamically sizing memory and might decrease GC pauses cased by resizing.
```
s/cased/caused/, but TBH I doubt resizing causes any GC pauses... starting with a smaller heap does mean you have a smaller young generation and more old GCs, but on the other hand those old GCs might be faster because the old gen is also smaller.

slaskawi replied:
Fixed `caused` thanks!

If I interpret the [CMS](https://docs.oracle.com/javase/8/docs/technotes/guides/vm/gctuning/cms.html) manual correctly:

> Concurrently resize the heap and prepare the support data structures for the next collection cycle, using one processor.

That means that the additional space will be available after full collection. So I guess this paragraph is still valid. However it's a good point - "Java Performance - The Definitive Guide" suggests using `Xmn` when making both `Xms` and `Xmx` equal. I'm adding a small note on that.

danberindei replied:
"concurrently" means there's no pause

slaskawi replied:
Yes but those resized heap will be available in the next collection cycle, so I interpret is after next full GC. 

Anyway - So do you propose to remove this whole fragment about `Xmx = Xms`? Or only part of it?

@alanfx - May I ask you for some input as well?

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+
+The most frequently used JVM flags are:
+
+* `XX:MaxGCPauseMillis` - Sets a target for the maximum GC pause time. Should be tuned to meet the SLA.
```
-XX

slaskawi replied:
Done

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+* `-XX:+ UseConcMarkSweepGC` - Enables usage of the CMS collector.
+* `-XX:+ CMSClassUnloadingEnabled` - Allows class unloading when the CMS collector is enabled.
+* `-XX:+ UseParNewGC` - Utilize a parallel collector for the young generation. This parameter minimizes pausing by using multiple collection threads in parallel.
+* `-XX:+ DisableExplicitGC` - Prevent explicit garbage collections.
```
No space after `+`

slaskawi replied:
Done

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+-XX:+UseLargePages
+-XX:+UseConcMarkSweepGC
+-XX:+UseParNewGC
+-XX:+DisableExplicitGC
```
Personally I'd remove the definitions of the JVM flags, and instead just provide an example of "expected" JVM configuration.

slaskawi replied:
Could you please elaborate more? What do you mean by "expected" JVM configuration? Can you share an example (which I could use in this manual)?

danberindei replied:
I meant the example you already have, but without the sections with the JVM flag definitions. I guess you could also use the  defaults from the server.

Perhaps removing all the definitions is a bit extreme, so maybe I'd keep a short explanation of the settings used in the example, but I wouldn't try to explain flags that we're not using.

slaskawi replied:
Hmmm I still do not quite understand. 

All flags have been described in previous paragraphs (like `Garbage collection` or `Memory settings`).

I'm also not sure what JVM flag definitions you have in mind and what defaults from the server (do you mean Infinispan server?).

Maybe if you could give me an example here it would be crystal clear.

danberindei replied:
Yes, you have described all the flags in the example, but also other flags, like `-XX:+CMSClassUnloadingEnabled`, and because they're not in the example it's not clear for users whether they should add it. That's why I'd like to start with our example/default JVM settings, and only explain those settings that we actually use. E.g.

    == JVM configuration

    This is an example configuration for a JVM with 32GB of heap:

        -server
        -Xmx32G

    * `-server`: select server mode, which enables the `C2` compiler. Already enabled by default on Linux.
    * `-Xmx32G`: set the maximum heap size.

    For more information, see the http://www.oracle.com/technetwork/articles/java/vmoptions-jsp-140102.html

And yes, by defaults I meant the defaults from our Infinispan Server startup scripts. Unfortunately I can't find anything that doesn't use `-Xmx512m` ATM...

slaskawi replied:
Ok got it now, thanks for being patient ;)

Maybe let's split this into two pieces - for server mode use settings similar to this and for library use this...

Let me write it in the docs. If it still won't fit your expectations we'll refactor it further...

slaskawi replied:
Argh... It probably won't be that easy because if we squashed all Java sub-paragraphs (`Memory settings`, `Garbage collection` and `Other settings`) we will lose some background information like this:

> The main goal is to minimize the amount of time when JVM is paused. Having said that, CMS is a suggested GC for Infinispan applications.

or this:

> Using `Xms` equal to `Xmx` will prevent JVM from dynamically sizing memory and might decrease GC pauses caused by resizing. 
Also when using this combination, it is advices to specify the `Xmn` parameter as well. 

So the question is, do we still want to do it?

@alanfx - Would be great to hear your input here also.

slaskawi replied:
@danberindei, @alanfx Thoughts?

alanfx replied:
@slaskawi I did comment about about setting `-Xmx` and `-Xms` to the same value. I think it is fine to keep that in the docs. I do wonder about changing to use the G1 collector by default, but that will require changing the server startup scripts as well and diverge from Wildfly. Not sure we want to do that yet, but G1 did perform better than CMS in my last round of testing.

danberindei replied:
Ok to keep it as is, and add `-XX:+UseG1GC` as an option alongside CMS.

slaskawi replied:
I added another example configuration with G1. I hope that's fine...

slaskawi replied:
@alanfx Maybe we should say something like this... In any case please use CMS but if your JDK is greater than version X (could you please tell me which JDK it was?), than use G1.

slaskawi replied:
I added configuration example (with G1) and a note that in the latest JVMs this GC might perform slightly better.

tristantarrant replied:
I think we really need to revise our startup scripts for server, since they are sized far too large for a "starter" configuration.

alanfx replied:
@slaskawi I think mentioning the "latest JVMs" is a good option. I ran my tests with JVM version 1.8.0.91, but I'm not sure that's worth mentioning.

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+
+== Network configuration
+
+Infinispan uses TCP/IP for sending packets over the network.
```
Or UDP, if you configure the UDP stack ;)

slaskawi replied:
Yes. Also this is the case when using client/server mode. Added note about that.

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+
+=== Ignore return values
+
+When you're not interested in returning value by `#put` or `#remove` method, use `Flag.IGNORE_RETURN_VALUES` flag as shown below:
```
s/in returning value by/in the return value of the/
Also, I'd rather use `Cache.put(k, v)` in the docs instead of `#put`.

slaskawi replied:
Done

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+[source,java]
+.Flag.IGNORE_RETURN_VALUES
+----
+cache.getAdvancedCache().withFlags(Flag.IGNORE_RETURN_VALUES)
```
Looks a bit incomplete...

    Cache noPreviousValueCache = cache.getAdvancedCache().withFlags(Flag.IGNORE_RETURN_VALUES);
    noPreviousValueCache.put(k, v);

slaskawi replied:
Done

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+
+=== Skip cache load
+
+When you don't want to load an entry from a Cache Store, use `Flag.SKIP_CACHE_LOAD` as shown below:
```
I wouldn't mention `SKIP_CACHE_LOAD` here. It's already covered by `IGNORE_RETURN_VALUES` for `put`, and for `get` you have to really know what you're doing in order to use it.

slaskawi replied:
@rvansa also mentioned that it can break stuff. I agree - let's remove it.

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+
+=== Use `Externalizable` for marshalling
+
+Infinispan uses JBoss Marshalling to transfer objects over the wire. The most efficient way to marshall user data is to provide an Externalizer:
```
I'd say more about the advantage of **`AdvancedExternalizer`** vs `Serializable`/`Externalizable`: for `Serializable`/`Externalizable`, JBoss Marshalling always sends the name of the class over the wire, but with `AdvancedExternalizer` it only sends the externalizer id. AFAIK our non-advanced `Externalizer` behaves almost exactly like `Externalizable`.

The disadvantage of `AdvancedExternalizer` is that it duplicates objects that appear twice in the object graph, and doesn't work at all the object graph has cycles -- you have to use `InstanceReusingAdvancedExternalizer` instead. Of course, it's always better if you can serialize just the root of the graph and have it write its children as primitives.

slaskawi replied:
That's a lot of very useful information :) I'll try to summarize that and hide some of the details you mentioned (I wouldn't like to mention too much of our internal behavior since it can get out of date very quickly)

danberindei commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+
+When the majority of stored data are Strings there are additional optimizations which can be taken into consideration:
+
+* Using `String#getBytes("UTF-8")` allows to decrease size of the object
```
I'd rephrase this as "If your strings are mostly ASCII, convert them to `UTF-8` and store them as `byte[]`. Note that the `UTF-8` version can be bigger for non-ASCII strings."

slaskawi replied:
Done

tristantarrant commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+
+== Capacity planning
+
+Infinispan stores data in memory. Depending on operating mode data is stored in serialized form (server mode) or directly
```
"on the operating mode"

tristantarrant replied:
Actually this needs a little more clarification:
Data in Infinispan is either stored as plain Java objects or in a serialized form, depending on operating mode (embedded or server) or on specific configuration options (store-as-binary).

slaskawi replied:
Done

tristantarrant commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,232 @@
+== Capacity planning
+
+Infinispan stores data in memory. Depending on operating mode data is stored in serialized form (server mode) or directly
+directly in memory (library mode). Data size can be estimated using sophisticated tools like  http://openjdk.java.net/projects/code-tools/jol/[Java Object Layout]
```
"directly" repeated

tristantarrant replied:
Also you provide a way to estimate the size of a POJO using JOL, but this doesn't provide information about serialized size, which depends on the marshaller. It would be nice to have a similar tool

slaskawi replied:
That would be very nice.... However I have no idea whether such tool exists... Maybe @galderz could help?

gustavonalle replied:
Minimally you just need to:
```
GenericJBossMarshaller marshaller = new GenericJBossMarshaller();
ByteBuffer byteBuffer = marshaller.objectToBuffer( /* my custom object */);
int size = byteBuffer.getLength();
```

alanfx commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,224 @@
+* `-Xmn` - Defines the minimum and maximum value for the young generation.
+* `-XX:NewRatio` - Define the ratio between young and old generations. Should not be used if -Xmn is enabled.
+
+Using `Xms` equal to `Xmx` will prevent JVM from dynamically sizing memory and might decrease GC pauses caused by resizing. 
```
This is the only thing I see in the Oracle docs about setting `Xms` equal to `Xmx`: 

> Setting -Xms and -Xmx to the same value increases predictability by removing the most important sizing decision from the virtual machine. However, the virtual machine is then unable to compensate if you make a poor choice.

From: https://docs.oracle.com/javase/8/docs/technotes/guides/vm/gctuning/sizing.html#sthref22

I don't think this really implies that it improves performance significantly, just that it avoids work in JVM.

mgencur replied:
I remember the old times when we were testing that. And using same value for -Xms and -Xmx really improved performance significantly. I don't have the numbers though.

slaskawi replied:
@alanfx @mgencur So is it OK to leave it as is?

alanfx replied:
@slaskawi yes, I think it is ok to leave it as is

danberindei replied:
ok

galderz replied:
Yeah, I'm not sure whether providing the same amount on `Xms` and `Xmx` makes any difference these days. We should test this.

slaskawi replied:
@galderz I believe it does, see: https://github.com/infinispan/infinispan/pull/4540#r79192021

slaskawi commented:
Updated, ready for the next round.

galderz commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,227 @@
+
+The most frequently used JVM flags are:
+
+* `-XX:MaxGCPauseMillis` - Sets a target for the maximum GC pause time. Should be tuned to meet the SLA.
```
Does this option really work?

alanfx replied:
In the testing I did, this option did have some positive effect on the throughput, but it is a "goal" and not a "hard limit". The GC will try to meet the goal if it can, but there are no guarantees.

galderz commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,227 @@
+
+* `-XX:MaxGCPauseMillis` - Sets a target for the maximum GC pause time. Should be tuned to meet the SLA.
+* `-XX:+UseConcMarkSweepGC` - Enables usage of the CMS collector.
+* `-XX:+CMSClassUnloadingEnabled` - Allows class unloading when the CMS collector is enabled.
```
Is this option more designed for managed environments when there are deploy/undeploy cycles?

slaskawi replied:
I think a very good explanation is here: https://stackoverflow.com/questions/3334911/what-does-jvm-flag-cmsclassunloadingenabled-actually-do

In short - I think it should stay because of classes created on runtime using dynamic languages like Groovy. It also won't hurt any traditional app...

galderz commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,227 @@
+noPreviousValueCache.put(k, v);
+----
+
+=== Use `Externalizable` for marshalling
```
You mean `Externalizer`?

slaskawi replied:
Good point! Fixed, thanks!

danberindei replied:
Actually, it should be `AdvancedExternalizer` :)

galderz commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,227 @@
+GlobalConfigurationBuilder builder = ...
+builder.serialization().addAdvancedExternalizer(new Book.BookExternalizer());
+----
+
```
Maybe you could add a link to the docu on externalizers if the reader is interested in more details?

slaskawi replied:
Done

slaskawi commented:
Updated, ready for the next round.

tristantarrant commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,227 @@
+
+== Capacity planning
+
+Data in Infinispan is either stored as plain Java objects or in a serialized form, depending on operating mode (embedded or server) or on specific configuration options (store-as-binary). 
```
Link to the user guide: ../user_guide/user_guide.html#store_as_binary

slaskawi replied:
Done

tristantarrant commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,227 @@
+
+Data in Infinispan is either stored as plain Java objects or in a serialized form, depending on operating mode (embedded or server) or on specific configuration options (store-as-binary). 
+Data size can be estimated using sophisticated tools like  http://openjdk.java.net/projects/code-tools/jol/[Java Object Layout]
+and total amount of required memory can be roughly estimated using the following formulas:
```
and *the* total amount of memory

tristantarrant commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,241 @@
+* `-XX:NewRatio` - Define the ratio between young and old generations. Should not be used if -Xmn is enabled.
+
+Using `Xms` equal to `Xmx` will prevent JVM from dynamically sizing memory and might decrease GC pauses caused by resizing. 
+Also when using this combination, it is advices to specify the `Xmn` parameter as well. 
```
it is **advised** to specify

Also, I'd explain why specifying Xmn is beneficial (generally because a lot of the allocations have a very short lifespan, and during high loads we generate a lot of them which, with a small NewSize would erroneously get promoted to older generations)

slaskawi replied:
I replaces the last sentence with this:

> It is a good practice to specify `Xmn` parameter. This guaranteed proper behavior during load peak (in such case Infinispan generates lots of small, short living objects).

I hope this will do the trick.

tristantarrant commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,241 @@
+----
+Cache noPreviousValueCache = cache.getAdvancedCache().withFlags(Flag.IGNORE_RETURN_VALUES);
+noPreviousValueCache.put(k, v);
+----
```
Should we mention that this flag can be turned on everywhere using the **unsafe** configuration options ?

tristantarrant commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,241 @@
+builder.serialization().addAdvancedExternalizer(new Book.BookExternalizer());
+----
+
+For more information, please refer to the http://infinispan.org/docs/dev/user_guide/user_guide.html#benefits_of_externalizers[User's Guide].
```
Don't use absolute links, but ones relative to this doc, so they always reference the correct version

slaskawi commented:
@tristantarrant I addressed your comments and updated the PR.

Other changes introduced:
* Removed TODO entry from documentation mini sprint
* Referenced performance guide from User Guide as well as FAQ

tristantarrant commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,250 @@
+builder.serialization().addAdvancedExternalizer(new Book.BookExternalizer());
+----
+
+For more information, please refer to the http://infinispan.org/docs/{infinispanversion}/user_guide/user_guide.html#benefits_of_externalizers[User's Guide].
```
Here you should use a relative link:
link:$$../user_guide/user_guide.html$$[User's Guide]

tristantarrant commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,271 @@
+[stem]
+.Required memory in Distributed mode
+++++
+Required Memo ry = Total Data Set*(Node Failures + 2)/(Nodes - Node Failures)
```
Extra space in the word 'Memory'

tristantarrant replied:
It looks like this is rendered using MathJax in the browser (which looks fine) but is blank in the PDF...

tristantarrant commented on documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc:
```diff
@@ -0,0 +1,271 @@
+
+Calculated amount of memory should be used for setting `Xmx` and `Xms` parameters.
+
+JVM as well as Infinispan require additional memory for other tasks like searches, allocating network buffers etc.
```
Better write it as: "Infinispan will also require additional memory..."

tristantarrant replied:
You missed this

tristantarrant commented:
Pushed to master, thanks !
